File: apps/mobile-app/utils/db/mappers/ItemMapper.ts
import type { Item, ItemField, ItemTagRef, ItemType } from '@/utils/dist/core/models/vault';

/**
 * Item with optional DeletedAt field for recently deleted items.
 */
export type ItemWithDeletedAt = Item & { DeletedAt?: string };

/**
 * Raw item row from database query.
 */
export type ItemRow = {
  Id: string;
  Name: string;
  ItemType: string;
  FolderId: string | null;
  FolderPath: string | null;
  Logo: Uint8Array | null;
  HasPasskey: number;
  HasAttachment: number;
  HasTotp: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
};

/**
 * Raw tag row from database query.
 */
export type TagRow = {
  ItemId: string;
  Id: string;
  Name: string;
  Color: string | null;
};

/**
 * Mapper class for converting database rows to Item objects.
 */
export class ItemMapper {
  /**
   * Map a single database row to an Item object.
   * @param row - Raw item row from database
   * @param fields - Processed fields for this item
   * @param tags - Tags for this item
   * @returns Item object
   */
  public static mapRow(
    row: ItemRow,
    fields: ItemField[] = [],
    tags: ItemTagRef[] = []
  ): Item {
    return {
      Id: row.Id,
      Name: row.Name,
      ItemType: row.ItemType as ItemType,
      Logo: row.Logo ?? undefined,
      FolderId: row.FolderId,
      FolderPath: row.FolderPath || null,
      Tags: tags,
      Fields: fields,
      HasPasskey: row.HasPasskey === 1,
      HasAttachment: row.HasAttachment === 1,
      HasTotp: row.HasTotp === 1,
      CreatedAt: row.CreatedAt,
      UpdatedAt: row.UpdatedAt
    };
  }

  /**
   * Map multiple database rows to Item objects with their fields and tags.
   * @param rows - Raw item rows from database
   * @param fieldsByItem - Map of ItemId to array of fields
   * @param tagsByItem - Map of ItemId to array of tags
   * @returns Array of Item objects
   */
  public static mapRows(
    rows: ItemRow[],
    fieldsByItem: Map<string, ItemField[]>,
    tagsByItem: Map<string, ItemTagRef[]>
  ): Item[] {
    return rows.map(row => this.mapRow(
      row,
      fieldsByItem.get(row.Id) || [],
      tagsByItem.get(row.Id) || []
    ));
  }

  /**
   * Group tag rows by ItemId into a map.
   * @param tagRows - Raw tag rows from database
   * @returns Map of ItemId to array of ItemTagRef
   */
  public static groupTagsByItem(tagRows: TagRow[]): Map<string, ItemTagRef[]> {
    const tagsByItem = new Map<string, ItemTagRef[]>();

    for (const tag of tagRows) {
      if (!tagsByItem.has(tag.ItemId)) {
        tagsByItem.set(tag.ItemId, []);
      }
      tagsByItem.get(tag.ItemId)!.push({
        Id: tag.Id,
        Name: tag.Name,
        Color: tag.Color || undefined
      });
    }

    return tagsByItem;
  }

  /**
   * Map tag rows to ItemTagRef array (for single item).
   * @param tagRows - Raw tag rows without ItemId
   * @returns Array of ItemTagRef
   */
  public static mapTagRows(tagRows: Omit<TagRow, 'ItemId'>[]): ItemTagRef[] {
    return tagRows.map(tag => ({
      Id: tag.Id,
      Name: tag.Name,
      Color: tag.Color || undefined
    }));
  }

  /**
   * Map a single item row for recently deleted items (includes DeletedAt).
   * @param row - Raw item row with DeletedAt
   * @param fields - Processed fields for this item
   * @returns Item object with DeletedAt
   */
  public static mapDeletedItemRow(
    row: ItemRow & { DeletedAt: string },
    fields: ItemField[] = []
  ): ItemWithDeletedAt {
    return {
      Id: row.Id,
      Name: row.Name,
      ItemType: row.ItemType as ItemType,
      Logo: row.Logo ? new Uint8Array(row.Logo) : undefined,
      FolderId: row.FolderId,
      FolderPath: row.FolderPath,
      DeletedAt: row.DeletedAt,
      HasPasskey: row.HasPasskey === 1,
      HasAttachment: row.HasAttachment === 1,
      HasTotp: row.HasTotp === 1,
      Fields: fields,
      CreatedAt: row.CreatedAt,
      UpdatedAt: row.UpdatedAt
    };
  }
}

File: apps/mobile-app/ios/VaultStoreKit/Database/Mappers/ItemMapper.swift
import Foundation
import SQLite
import VaultModels
import VaultUtils

/// Raw item row from database query.
public struct ItemRow {
    public let id: String
    public let name: String?
    public let itemType: String
    public let folderId: String?
    public let folderPath: String?
    public let logo: Data?
    public let hasPasskey: Bool
    public let hasAttachment: Bool
    public let hasTotp: Bool
    public let createdAt: String
    public let updatedAt: String
    public let deletedAt: String?

    /// Initialize from a database row dictionary.
    public init?(from row: [String: Any]) {
        guard let id = row["Id"] as? String,
              let itemType = row["ItemType"] as? String,
              let createdAt = row["CreatedAt"] as? String,
              let updatedAt = row["UpdatedAt"] as? String else {
            return nil
        }

        self.id = id
        self.name = row["Name"] as? String
        self.itemType = itemType
        self.folderId = row["FolderId"] as? String
        self.folderPath = row["FolderPath"] as? String

        // Handle logo data - can be base64 string or Blob
        if let logoBase64 = row["Logo"] as? String {
            self.logo = Data(base64Encoded: logoBase64)
        } else if let logoBlob = row["Logo"] as? SQLite.Blob {
            self.logo = Data(logoBlob.bytes)
        } else {
            self.logo = nil
        }

        self.hasPasskey = (row["HasPasskey"] as? Int64 ?? 0) == 1
        self.hasAttachment = (row["HasAttachment"] as? Int64 ?? 0) == 1
        self.hasTotp = (row["HasTotp"] as? Int64 ?? 0) == 1
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = row["DeletedAt"] as? String
    }
}

/// Raw tag row from database query.
public struct TagRow {
    public let itemId: String
    public let id: String
    public let name: String
    public let color: String?

    /// Initialize from a database row dictionary.
    public init?(from row: [String: Any]) {
        guard let itemId = row["ItemId"] as? String,
              let id = row["Id"] as? String,
              let name = row["Name"] as? String else {
            return nil
        }

        self.itemId = itemId
        self.id = id
        self.name = name
        self.color = row["Color"] as? String
    }
}

/// Tag reference for single item queries (without ItemId).
public struct SingleItemTagRow {
    public let id: String
    public let name: String
    public let color: String?

    /// Initialize from a database row dictionary.
    public init?(from row: [String: Any]) {
        guard let id = row["Id"] as? String,
              let name = row["Name"] as? String else {
            return nil
        }

        self.id = id
        self.name = name
        self.color = row["Color"] as? String
    }
}

/// Mapper class for converting database rows to Item objects.
public struct ItemMapper {
    /// Map a single database row to an Item object.
    /// - Parameters:
    ///   - row: Raw item row from database
    ///   - fields: Processed fields for this item
    /// - Returns: Item object
    public static func mapRow(_ row: ItemRow, fields: [ItemField] = []) -> Item? {
        guard let createdAt = DateHelpers.parseDateString(row.createdAt),
              let updatedAt = DateHelpers.parseDateString(row.updatedAt) else {
            return nil
        }

        return Item(
            id: UUID(uuidString: row.id) ?? UUID(),
            name: row.name,
            itemType: row.itemType,
            logo: row.logo,
            folderId: row.folderId.flatMap { UUID(uuidString: $0) },
            folderPath: row.folderPath,
            fields: fields,
            hasPasskey: row.hasPasskey,
            hasAttachment: row.hasAttachment,
            hasTotp: row.hasTotp,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }

    /// Map multiple database rows to Item objects with their fields.
    /// - Parameters:
    ///   - rows: Raw item rows from database
    ///   - fieldsByItem: Dictionary of ItemId to array of fields
    /// - Returns: Array of Item objects
    public static func mapRows(
        _ rows: [ItemRow],
        fieldsByItem: [String: [ItemField]]
    ) -> [Item] {
        return rows.compactMap { row in
            let fields = fieldsByItem[row.id] ?? []
            return mapRow(row, fields: fields)
        }
    }

    /// Map a single item row for recently deleted items (includes DeletedAt).
    /// - Parameters:
    ///   - row: Raw item row with DeletedAt
    ///   - fields: Processed fields for this item
    /// - Returns: Item object (deletedAt stored as extension or separate property if needed)
    public static func mapDeletedItemRow(_ row: ItemRow, fields: [ItemField] = []) -> Item? {
        return mapRow(row, fields: fields)
    }
}
